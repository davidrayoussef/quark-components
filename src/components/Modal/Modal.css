.overlay {
  position: fixed;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100vw;
  height: 100vh;
  background: rgba(0, 0, 0, .3);
  z-index: 1;
}

.modal {
  position: relative;
  display: grid;
  grid-gap: 35px;
  align-content: center;
  justify-content: center;
  padding: 50px;
  width: 50%;
  height: auto;
  background: #FFF;
  text-align: center;
}

.modal header {
  font-size: 2rem;
  font-weight: 100;
}

.x-button {
  position: absolute;
  top: 25px;
  right: 25px;
  cursor: pointer;
  width: 20px;
  fill: rgba(0, 0, 0, .3);
  stroke: rgba(0, 0, 0, .3);
  stroke-width: 2;
  transition: .2s;
}

.x-button:hover {
  fill: rgba(0, 0, 0, .8);
  stroke: rgba(0, 0, 0, .8);
}

@media (max-width: 800px) {
  .modal {
    width: 100vw;
  }
}
