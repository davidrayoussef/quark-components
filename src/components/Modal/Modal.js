import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from '../Button/Button';
import style from './Modal.css';

class Modal extends Component {
  state = {
    isOpen: false
  };

  static proptypes = {
    children: PropTypes.node,
    title: PropTypes.string,
    content: PropTypes.string
  };

  static getDerivedStateFromProps(nextProps, prevState, prevProps) {
    if (prevProps === null || nextProps.isOpen !== prevState.isOpen) {
      return {
        isOpen: nextProps.isOpen
      };
    }
    return null;
  };

  handleClose = () => {
    this.setState({
      isOpen: false
    });
  };

  handleEsc = (e) => {
    console.log(e.which);
  };

  render() {
    return this.state.isOpen ? (
      <div className={style.overlay}>
        <div className={style.modal}>
          <header>{this.props.title}</header>
          {this.props.children}
          <div className={style.buttonContainer}>
            <span onClick={this.handleClose} onKeyUp={this.handleEsc}>
              <Button type="secondary">OK</Button>
            </span>
          </div>
        </div>
      </div>
    ) : null;
  }
};

export default Modal;
