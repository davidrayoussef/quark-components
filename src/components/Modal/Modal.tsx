import React, { Component } from 'react';

import { Button } from '../Button';
import { Icon } from '../Icon';

import { ButtonType, ModalProps } from '../../shared';

import style from './Modal.css';

export class Modal extends Component<ModalProps> {
  modalElement: React.RefObject<HTMLDivElement> = React.createRef();

  componentDidMount(): void {
    window.addEventListener('keyup', this.handleEsc);
  }

  componentWillUnMount(): void {
    window.removeEventListener('keyup', this.handleEsc);
  }

  handleEsc = (event: KeyboardEvent): void => {
    if (event.key === 'Escape') {
      this.props.handleClose();
    }
  };

  handleClickOutside = (event: React.MouseEvent<HTMLDivElement>): void => {
    if (!this.modalElement.current.contains(event.target as Node)) {
      this.props.handleClose();
    }
  };

  render(): JSX.Element {
    const { isOpen, title, children, handleClose } = this.props;
    return isOpen ? (
      <div className={style.overlay} onClick={this.handleClickOutside}>
        <div className={style.modal} ref={this.modalElement}>
          <Icon
            value="close"
            width="20px"
            color="rgba(0, 0, 0, .3)"
            className={style.xButton}
            onClick={handleClose}
          />
          <header>{title}</header>
          {children}
          <Button
            styleType={ButtonType.Secondary}
            style={{ width: 100, margin: 'auto' }}
            onClick={handleClose}
          >
            OK
          </Button>
        </div>
      </div>
    ) : null;
  }
}
