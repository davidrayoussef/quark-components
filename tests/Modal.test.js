import React from 'react';
import { shallow, mount } from 'enzyme';
import Modal from '../src/components/Modal';

test('should render', () => {
  const wrapper = shallow(
    <Modal isOpen={true} handleClose={jest.fn}></Modal>
  );
  expect(wrapper);
});

test('should render correct title', () => {
  const wrapper = shallow(
    <Modal 
      title="Modal Title" 
      isOpen={true} 
      handleClose={jest.fn}>
    </Modal>
  );
  expect(wrapper.find('header').text()).toBe('Modal Title');
});

test('should render correct content', () => {
  const wrapper = shallow(
    <Modal isOpen={true} handleClose={jest.fn}>
      <p>Modal Content</p>
    </Modal>
  );
  expect(wrapper.find('p').text()).toBe('Modal Content');
});

test('should not render children if isOpen prop set to false', () => {
  const wrapper = shallow(
    <Modal isOpen={false} handleClose={jest.fn}>
      <p>Modal Content</p>
    </Modal>
  );
  expect(wrapper.children().length).toBe(0);
});

test('snapshot', () => {
  const wrapper = shallow(
    <Modal isOpen={true} handleClose={jest.fn}>
      <p>Modal Content</p>
    </Modal>
  );
  expect(wrapper).toMatchSnapshot();
});
