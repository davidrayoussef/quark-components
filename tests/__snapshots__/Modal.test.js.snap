// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="overlay"
  onClick={[Function]}
>
  <div
    className="modal"
  >
    <Icon
      className="x-button"
      color="rgba(0, 0, 0, .3)"
      onClick={[Function]}
      value="close"
      width="20px"
    />
    <header />
    <p>
      Modal Content
    </p>
    <Button
      onClick={[Function]}
      style={
        Object {
          "margin": "auto",
          "width": 100,
        }
      }
      type="secondary"
    >
      OK
    </Button>
  </div>
</div>
`;
